ovirt_imageio/client/_options.py

```diff
@@ -21,36 +23,75 @@ def __init__(self):
         self._parser = argparse.ArgumentParser(
             description="Transfer disk images")
         self._parser.set_defaults(
-            command=lambda _: self._parser.print_help())
+            command=lambda _: self._parser.print_help(),
+            password=None)
         self._commands = self._parser.add_subparsers(title="commands")
 
     def add_sub_command(self, name, help, func):
         cmd = self._commands.add_parser(name, help=help)
         cmd.set_defaults(command=func)
 
+        cmd.add_argument(
+            "-c", "--config",
+            help="If set, read specified section from the configuration "
+                 f"file ({self.config_file}).")
+
         cmd.add_argument(
             "--engine-url",
-            help="ovirt-engine URL.")
+            help="ovirt-engine URL. If not set, read from the specified "
+                 "config section (required).")
 
         cmd.add_argument(
             "--username",
-            help="ovirt-engine username.")
+            help="ovirt-engine username. If not set, read from the specified "
+                 "config section (required).")
 
         cmd.add_argument(
             "--password-file",
-            help="Read ovirt-engine password from file.")
+            help="Read ovirt-engine password from file. If not set, read "
+                 "password from the specified config section.")
 
         cmd.add_argument(
             "--cafile",
-            help="Path to ovirt-engine CA certificate")
+            help="Path to ovirt-engine CA certificate. If not set, read from "
+                 "the specified config section")
 
         return cmd
 
-    def parse(self):
-        args = self._parser.parse_args()
-        # XXX Read config file and merge into args.
+    def parse(self, args=None):
+        args = self._parser.parse_args(args=args)
+        if args.config:
+            self._merge_config(args)
+
+        for name in ("engine_url", "username"):
+            if getattr(args, name, None) is None:
+                self._parser.error(f"Option '{name}' is required")
+
         return args
 
+    @property
+    def config_file(self):
+        # https://specifications.freedesktop.org/basedir-spec
+        base_dir = os.environ.get("XDG_CONFIG_HOME")
+        if not base_dir:
+            base_dir = os.path.expanduser("~/.config")
+        return os.path.join(base_dir, "ovirt-img.conf")
+
+    def _merge_config(self, args):
+        config = configparser.ConfigParser(interpolation=None)
+        config.read([self.config_file])
+
+        if not config.has_section(args.config):
+            self._parser.error(
+                f"No section: '{args.config}' in '{self.config_file}'")
+
+        for name in ("engine_url", "username", "password", "cafile"):
```

Comment:
Ok, the "password/password-file"-thing seems to be a feature of the parser? Because it's the same for "engine/engine-url".

Reply:
When you add --foo-bar the parser save the value in args.foo_bar. You can specify dest= if you want to make this explicit but this is well known.

password and --password-file are not the same - there are 3 ways to specify a password:
- --password-file: read password from file
- password in config file: read password from config file
- not set: read from stdin

Reply:
And there's this behaviour, which to me seems unspecified, where it is trying to open a password file that was given with --password:

`
[rob@host-vdsm ovirt-imageio]$ ./ovirt-img download-disk --password x --engine-url https://rob-ovirt-engine.local --username admin@internal --cafile /home/rob/cert.pem -f raw f931f3b6-669a-46d2-8309-f23d0575cdaf  diskimage 
Traceback (most recent call last):
  File "./ovirt-img", line 13, in <module>
    _tool.main()
  File "/home/rob/share/git/nir/ovirt-imageio/ovirt_imageio/client/_tool.py", line 26, in main
    args.command(args)
  File "/home/rob/share/git/nir/ovirt-imageio/ovirt_imageio/client/_download.py", line 62, in download_disk
    con = _ovirt.connect(args)
  File "/home/rob/share/git/nir/ovirt-imageio/ovirt_imageio/client/_ovirt.py", line 46, in connect
    with open(args.password_file) as f:
FileNotFoundError: [Errno 2] No such file or directory: 'x'
`

Reply:
That behaviour might be "well known" to python-args-parser-programmers, but not necessarily to the users of this tool.

Reply:
This is a feature of argparse (and geopt) - you can use partial names when setting arguments. All of these are the same:

    --password-file
    --password
    --pass
    --p

If we had a --password argument this would not work, but we don't.

Reply:
Ok, to clarify: it seems to just automatically make short-options for the given long-options, as long as they are unique.